scsi_devfs_scan.8
.TH SG_DEVFS_SCAN "8" "April 2003" "sg3_utils-1.03" SG3_UTILS
.SH NAME
sg_devfs_scan \- Scan SCSI devices within a devfs tree
.SH SYNOPSIS
.B sg_devfs_scan
[\fI-d <dir>\fR] 
[\fI-i\fR] 
[\fI-ide\fR] 
[\fI-l [-x]\fR]
[\fI-q\fR]
<\fIgeneric device\fR>
.SH DESCRIPTION
.\" Add any additional description here
.PP
scsi_devfs_scan is a utility for doing a directory scan on a system
running devfs to identify SCSI (and optionally IDE) devices. Various
information (including an INQUIRY) can be listed for each found device.
.TP
-d <dir>
location of devfs [default: /dev ]
.TP
-i
show INQUIRY data for each SCSI device
.TP
-ide
show scan of IDE devices after SCSI ones
.TP
-l
show device file names in leaf directory
.TP
-x
add (major,minor) information to '-l'
.TP
-q
just output host, bus, target, lun numbers
.SH AUTHORS
Written by D. Gilbert
.SH COPYRIGHT
Copyright \(co 2001 Douglas Gilbert
.br
This software is distributed under the GPL version 2. There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


CHANGELOG
Each utility has its own version number, date of last change and
some description at the top of its ".c" file. All utilities in the main
directory have their own "man" pages. There is also a sg3_utils man page.

Changelog for sg3_utils-1.25 [20070629]
  - sg_dd: add oflag=sparse to step over bs*bpt number of zeros;
    - with oflag=sparse, write last bs*bpt segment at end or after
      error so file length of OFILE is appropriate
    - allow iflag=null and oflag=null which do nothing (placeholders)
    - when coe>1 then SCSI READ LONG logic remember extended block
      length of first encountered error
  - sg_ses: sync with ses2r17

Changelog for sg3_utils-1.24 [20070507]
  - sg_raw: new utility to send arbirary SCSI commands
  - sg_luns: increase number of luns that can be fetched
    - fix length of raw and hex output
    - add '--quiet' option to output only ASCII hex
      representation of each lun
  - sg_rtpg: update for changes in spc4r09
  - sg_persist: update documentation, spc-4 references
    - fix exit status values
  - sg_inq: update version descriptors per spc4r09
    - fix '--id' and '--extended'
    - extend block limits VPD page (sbc3r09)
  - sg_vpd: extend block limits VPD page (sbc3r09)
    - append relative target port identifier to SAS target
      port address with '-iq' option
  - sg_logs: add decoding for stats+performance log pages
    - fix showing of page names for pdt > 0
    - implement '-HH' for single and all pages, fix '-r'
    - when '--maxlen=' given, only do single fetch
    - add Tape Alert (ssc), Media and Element statistics (smc) pages 
    - add '--brief' option
  - sg_ses: sync with ses2r16
    - fix bay number for SAS
  - sg_format: add '--dcrt' and '--security' options
  - sgm_dd: add 'smmap' oflag for shared_mmap_io testing
    - add 'dio' oflag
  - sg_dd, sgp_dd: add 'dio' iflag and oflag
  - sg_modes: change SAS mode page names per sas2r09
    - check validity of block descriptors length
  - sg_pt: change opaque context object from 'void *'
    to 'struct sg_pt_base *'
  - sg_opcodes: anticipate extra tmfs from 07-159r0
  - sg_sat_set_features: add more usage information
    - add man page
  - sg_sat_phy_event: add to examples directory
  - sg_lib: sync asc/ascq codes with spc4r10
  - Solaris port: using uscsi interface

Changelog for sg3_utils-1.23 [20070131]
  - sg_read_buffer: new utility
  - sg_write_buffer: new utility
  - sg_opcodes, sg_senddiag, sg_logs, sg_modes, sg_start, sg_inq,
    sg_turs, sg_readcap, sg_rbuf: add getopt_long() based cli;
    old and new cli selectable, new getopt_long cli is default
  - scripts: new subdirectory containing some bash scripts
    - add scripts/README file
  - sg_reassign: add '--hex' option for grown and primary lists
  - sg_rtpg: add '--raw' option
  - sg_lib.h, sg_cmds_basic.h + sg_cmds_extra.h: add C++ 
    'extern "C" ' wrappers
    - cleanup C code so it will compile as C++
  - sg_lib: sync with spc4r08
    - include <inttypes.h>, use PRId64 instead of %lld form
    - fix sg_get_sense_str() when empty sense buffer
  - win32 port: add Makefile.mingw + related support for MinGW
  - sg_cmds_extra: add sg_ll_read_buffer() and sg_ll_write_buffer()
  - sg_dd, sgp_dd, sgm_dd, sg_read: use lseek64() instead of llseek.c
  - sgm_dd: accept coe=<n> for interworking with sg_dd
  - sg_rdac: fix on non-linux ports
  - sg_ses: fix spurious warning in additional element status page
    - '-rr' option outputs a diagnostic page in binary to stdout
  - sg_opcodes: add command timeout descriptor support (spc4r08)
    - change linux specific pass through to generic pass through
  - sg_logs: add 'name=value' decoding for SAS specific lpage
  - examples+utils subdirectories: remove symlinks
  - synchronize man pages with usage messages
  - sg3_utils.spec: rework

Changelog for sg3_utils-1.22 [20061016]
  - sgp_dd: accept verbose=<n> as well as deb=<n> to ease
    interworking with sg_dd and sgm_dd
  - sg_sat_set_features: added to examples directory
  - sg_lib: sync asc/ascq text with spc4r06
    - move SG_LIB_CAT_NO_SENSE and SG_LIB_CAT_RECOVERED to
      20 and 21 respectively; add SG_LIB_CAT_ABORTED_COMMAND
      at 11 (its sense key value) 
  - sg_vpd: tweak '--page=sp --quiet' output
    - change '-HHH' so same as '-rr' (prepares ATA Information
      (ai) response for hdparm)
  - sg_requests: add '-s' option to set exit status from
    parameter data
  - sg_modes: exit quickly from '-e' if device not ready
  - sg_logs: sync sas log pages with sas2r05a
    - expand background scan results log page
    - add '-m=<max_len>' to limit response length
    - drop '-scum' and '-sthr' options and add '-select'
  - sg_write_long: add '--16' option to send 16 byte cdb
    - add '--wr_uncor' and '--pblock' options
  - sg_senddiag: cleanup and add sdiag_sas_p1_stop.txt
    to examples directory
  - sg_format: add '--cmplst=<n>' option (default: 1)
    - add '--pfu=<n>' option
    - expand man page to discuss P/D/C/GLISTs
  - sg_reassign: add '--primary' option to fetch primary
    defect list (PLIST) length
  - sg_readcap: add '-H' option to output response in hex
    and '-r' to output response in binary to stdout
    - add logical blocks per physical block (sbc3r07)
  - sginfo: add PLIST and GLIST designation to defect lists
  - sg_cmds: split this support file into sg_cmds_basic.[hc]
    and sg_cmds_extra.[hc]
    - add sg_ll_ata_pt() (SATL ATA pass) to sg_cmds_extra.[hc]
  - sg_rdac: fix includes for FreeBSD
  - sg_dd: add 'coe_limit=' option to exit after <n>
    consecutive 'coe' type read errors
  - sgm_dd: print out throughput information when signal arrives
    if time=1 (like sg_dd does already)
  - sg_inq: change '-HHH' so same as '-rr'. Now sg_inq, sg_vpd
    and sdparm output for hdparm with '-HHH'
    -add '-l=<resp_len>' option
  - sg_read_long: add '--pblock' option for physical blocks
  - sg_luns: add '--hex' and '--raw' options
  - sg_requests: add '--hex' and '--raw' options

archive/scsi_devfs_scan.c
#define _XOPEN_SOURCE 500
#define _GNU_SOURCE

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/sysmacros.h>
#include <sys/stat.h>
#include <sys/ioctl.h>
#include <dirent.h>
#include <errno.h>
#include <unistd.h>
#include "sg_lib.h"
#include "sg_io_linux.h"

/* Code for scanning for SCSI devices within a Linux device pseudo file
   system.
   *  Copyright (C) 2001 D. Gilbert
   *  This program is free software; you can redistribute it and/or modify
   *  it under the terms of the GNU General Public License as published by
   *  the Free Software Foundation; either version 2, or (at your option)
   *  any later version.

      This program scans the /dev directory structure looking for the
      devfs "primary" scsi (and optionally IDE) device names.

   Version 0.13 20030430
*/

void usage()
{
    printf("Usage: 'scsi_devfs_scan [-d <dir>] [-i] [-ide] [-l [-x]] "
           "[-q]'\n");
    printf("    where: -d <dir> location of devfs [default: /dev ]\n");
    printf("           -i   show INQUIRY data for each SCSI device\n");
    printf("           -ide show scan of IDE devices after SCSI devices\n");
    printf("           -l   show device file names in leaf directory\n");
    printf("           -q   just output host, bus, target, lun numbers\n");
    printf("           -x   add (major,minor) information to '-l'\n");
}

#define NAME_LEN_MAX 256
#define LEVELS 4

#define SENSE_BUFF_LEN 32       /* Arbitrary, could be larger */
#define DEF_TIMEOUT 60000       /* 60,000 millisecs == 60 seconds */
#define INQUIRY_CMD     0x12
#define INQUIRY_CMDLEN  6

static const char * level_arr[LEVELS] = {"host", "bus", "target", "lun"};

static int do_ide = 0;
static int do_inq = 0;
static int do_leaf = 0;
static int do_extra = 0;
static int do_quiet = 0;
static int checked_sg = 0;

static void ddStrHex(const char* str, int len)
{
    const char* p = str;
    unsigned char c;
    char buff[82];
    int a = 0;
    const int bpstart = 5;
    const int cpstart = 60;
    int cpos = cpstart;
    int bpos = bpstart;
    int i, k;

    if (len <= 0) return;
    memset(buff,' ',80);
    buff[80]='\0';
    k = sprintf(buff + 1, "%.2x", a);
    buff[k + 1] = ' ';
    if (bpos >= ((bpstart + (9 * 3))))
        bpos++;

    for(i = 0; i < len; i++)
    {
        c = *p++;
        bpos += 3;
        if (bpos == (bpstart + (9 * 3)))
            bpos++;
        sprintf(&buff[bpos], "%.2x", (int)(unsigned char)c);
        buff[bpos + 2] = ' ';
        if ((c < ' ') || (c >= 0x7f))
            c='.';
        buff[cpos++] = c;
        if (cpos > (cpstart+15))
        {
            printf("%s\n", buff);
            bpos = bpstart;
            cpos = cpstart;
            a += 16;
            memset(buff,' ',80);
            k = sprintf(buff + 1, "%.2x", a);
            buff[k + 1] = ' ';
        }
    }
    if (cpos > cpstart)
    {
        printf("%s\n", buff);
    }
}

static int do_inquiry(int sg_fd, void * resp, int mx_resp_len)
{
    int res;
    unsigned char inqCmdBlk[INQUIRY_CMDLEN] = {INQUIRY_CMD, 0, 0, 0, 0, 0};
    unsigned char sense_b[SENSE_BUFF_LEN];
    sg_io_hdr_t io_hdr;

    inqCmdBlk[4] = (unsigned char)mx_resp_len;
    memset(&io_hdr, 0, sizeof(sg_io_hdr_t));
    io_hdr.interface_id = 'S';
    io_hdr.cmd_len = sizeof(inqCmdBlk);
    io_hdr.mx_sb_len = sizeof(sense_b);
    io_hdr.dxfer_direction = SG_DXFER_TO_FROM_DEV;
    io_hdr.dxfer_len = mx_resp_len;
    io_hdr.dxferp = resp;
    io_hdr.cmdp = inqCmdBlk;
    io_hdr.timeout = DEF_TIMEOUT;

    if (ioctl(sg_fd, SG_IO, &io_hdr) < 0) {
        perror("SG_IO (inquiry) error");
        return -1;
    }
    res = sg_err_category3(&io_hdr);
    switch (res) {
    case SG_LIB_CAT_CLEAN:
    case SG_LIB_CAT_RECOVERED:
        return 0;
    default:
        sg_chk_n_print3("Failed INQUIRY", &io_hdr, 1);
        return -1;
    }
}

void leaf_dir(const char * lf, unsigned int * larr)
{
    char name[NAME_LEN_MAX * 2];
    int res;

    if (do_quiet) {
        printf("%u\t%u\t%u\t%u\n", larr[0], larr[1], larr[2], larr[3]);
        return;
    }
    printf("%u\t%u\t%u\t%u\t%s\n", larr[0], larr[1], larr[2], larr[3], lf);
    if (do_leaf) {
        struct dirent * de_entry;
        struct dirent * de_result;
        DIR * sdir;
        int outpos;

        if (NULL == (sdir = opendir(lf))) {
            fprintf(stderr, "leaf_dir: opendir of %s: failed\n", lf);
            return;
        }
        de_entry = (struct dirent *)malloc(sizeof(struct dirent) + 
                                           NAME_LEN_MAX);
        if (NULL == de_entry)
            return;
        res = 0;
        printf("\t");
        outpos = 8;
        while (1) {
            res = readdir_r(sdir, de_entry, &de_result);
            if (0 != res) {
                fprintf(stderr, "leaf_dir: readdir_r of %s: %s\n", 
                        lf, strerror(res));
                res = -2;
                break;
            }
            if (de_result == NULL) 
                break;
            strncpy(name, de_entry->d_name, NAME_LEN_MAX * 2);
            if ((0 == strcmp("..", name)) ||(0 == strcmp(".", name))) 
                continue;
            if (do_extra) {
                struct stat st;
                char devname[NAME_LEN_MAX * 2];

                strncpy(devname, lf, NAME_LEN_MAX * 2);
                strcat(devname, "/");
                strcat(devname, name);
                if (stat(devname, &st) < 0)
                    return;
                if (S_ISCHR(st.st_mode)) {
                    strcat(name, "(c ");
                    sprintf(name + strlen(name), "%d %d)", major(st.st_rdev),
                            minor(st.st_rdev));
                }
                else if (S_ISBLK(st.st_mode)) {
                    strcat(name, "(b ");
                    sprintf(name + strlen(name), "%d %d)", major(st.st_rdev),
                            minor(st.st_rdev));
                }
            }
            res = strlen(name);
            if ((outpos + res + 2) > 80) {
                printf("\n\t");
                outpos = 8;
            }
            printf("%s  ", name);
            outpos += res + 2;
        }
        printf("\n");
    }
    if (do_inq) {
        int sg_fd;
        char buff[64];

        memset(buff, 0, sizeof(buff));
        strncpy(name, lf, NAME_LEN_MAX * 2);
        strcat(name, "/generic");
        if ((sg_fd = open(name, O_RDONLY)) < 0) {
            if (! checked_sg) {
                checked_sg = 1;
                if ((sg_fd = open("/dev/sg0", O_RDONLY)) >= 0)
                    close(sg_fd);  /* try and get sg module loaded */
                sg_fd = open(name, O_RDONLY);
            }
            if (sg_fd < 0) {
                printf("Unable to open sg device: %s, %s\n", name, 
                       strerror(errno));
                return;
            }
        }
        if (0 != do_inquiry(sg_fd, buff, 64))
            return;
        close(sg_fd);
        ddStrHex(buff, 64);
    }
}

/* Return 0 -> ok, -1 -> opendir() error, -2 -> readdir_r error, 
         -3 -> malloc error */
int hbtl_scan(const char * path, int level, unsigned int *larr)
{
    struct dirent * de_entry;
    struct dirent * de_result;
    char new_path[NAME_LEN_MAX * 2];
    DIR * sdir;
    int res;
    size_t level_slen;

    level_slen = strlen(level_arr[level]);
    if (NULL == (sdir = opendir(path))) {
        fprintf(stderr, "hbtl_scan: opendir of %s: failed\n", path);
        return -1;
    }
    de_entry = (struct dirent *)malloc(sizeof(struct dirent) + NAME_LEN_MAX);
    if (NULL == de_entry)
        return -3;
    res = 0;
    while (1) {
        res = readdir_r(sdir, de_entry, &de_result);
        if (0 != res) {
            fprintf(stderr, "hbtl_scan: readdir_r of %s: %s\n", 
                    path, strerror(res));
            res = -2;
            break;
        }
        if (de_result == NULL) 
            break;
        if (0 == strncmp(level_arr[level], de_entry->d_name, level_slen)) {
            if (1 != sscanf(de_entry->d_name + level_slen, "%u", larr + level))
                larr[level] = UINT_MAX;
            strncpy(new_path, path, NAME_LEN_MAX * 2);
            strcat(new_path, "/");
            strcat(new_path, de_entry->d_name);
            if ((level + 1) < LEVELS) {
                res = hbtl_scan(new_path, level + 1, larr);
                if (res < 0)
                    break;
            }
            else 
                leaf_dir(new_path, larr);
        }
    }
    free(de_entry);
    closedir(sdir);
    return res;
}

#define D_ROOT_SZ 512


int main(int argc, char * argv[])
{
    int k, res;
    char ds_root[D_ROOT_SZ];
    char di_root[D_ROOT_SZ];
    unsigned int larr[LEVELS];
    struct stat st;

    strncpy(ds_root, "/dev", D_ROOT_SZ);
    for (k = 1; k < argc; ++k) {
        if (0 == strcmp("-ide", argv[k]))
            do_ide = 1;
        else if (0 == strcmp("-i", argv[k]))
            do_inq = 1;
        else if (0 == strcmp("-l", argv[k]))
            do_leaf = 1;
        else if (0 == strcmp("-x", argv[k]))
            do_extra = 1;
        else if (0 == strcmp("-q", argv[k]))
            do_quiet = 1;
        else if (0 == strncmp("-d", argv[k], 2)) {
            if (strlen(argv[k]) > 2)
                strncpy(ds_root, argv[k] + 2, D_ROOT_SZ);
            else if (++k < argc)
                strncpy(ds_root, argv[k], D_ROOT_SZ);
        }
        else if ((0 == strcmp("-?", argv[k])) ||
                 (0 == strncmp("-h", argv[k], 2))) {
            printf("Scan SCSI devices within a devfs tree\n\n");
            usage();
            return 1;
        }
        else if (*argv[k] == '-') {
            printf("Unknown switch: %s\n", argv[k]);
            usage();
            return 1;
        }
        else if (*argv[k] != '-') {
            printf("Unknown argument\n");
            usage();
            return 1;
        }
    }
    strncpy(di_root, ds_root, D_ROOT_SZ);
    strcat(di_root, "/.devfsd");
    if (stat(di_root, &st) < 0)
        printf("Didn't find %s so perhaps devfs is not present,"
                " continuing ...\n", di_root);
    strncpy(di_root, ds_root, D_ROOT_SZ);
    strcat(ds_root, "/scsi");
    strcat(di_root, "/ide");

    if (do_ide)
        printf("SCSI scan:\n");
    res = hbtl_scan(ds_root, 0, larr);
    if (res < 0)
        printf("main: scsi hbtl_scan res=%d\n", res);
    do_inq = 0;  /* won't try SCSI INQUIRY on IDE devices */
    if (do_ide) {
        printf("\nIDE scan:\n");
        res = hbtl_scan(di_root, 0, larr);
        if (res < 0)
            printf("main: ide hbtl_scan res=%d\n", res);
    }
    return 0;
}
